### PR TITLE
feat(core): A major improvement to the tox groups

### PR DESCRIPTION
**Pull request by Monsterovich**
default
- Added messages when user joins/leaves the group.
- Added messages when user changes the nickname.
- Fixed bugs with refreshing the list of users in group.
- More fixes for users with empty nicknames.
- More support for friend aliases in groups.
- Fixed bug: users with no nickname with have literally every line high-lightened.

---
This change is [Reviewable](https://reviewable.io/reviews/qtox/qtox/5469)

### CODE REVIEW

**Comment by sudden6** [comment]
___
*[src/widget/form/groupchatform.cpp, line 330 at r1](https://reviewable.io/reviews/qtox/qtox/5469#-LUND_7L5pDMIYT4BLjY:-LUND_7L5pDMIYT4BLjZ:b9pzftg) ([raw file](https://github.com/qtox/qtox/blob/48d8b3b6ceaa72eb9e45ff82e6bec893e8bc4081/src/widget/form/groupchatform.cpp#L330)):*
> ```cpp
>     for (const auto& peerPk : peers.keys()) {
>         const QString name = peers.value(peerPk);
>         // ignore weird issue: when user joins the group, the name is empty, then it's renamed to normal nickname (why?)
> ```

IIRC the Tox protocol allows empty nicknames and the nickname is only sent after joining, usually we display the public key instead of the empty string,
Not sure if it would be better to display two messages: 1. " 123FABC... joined" and then a message like "123FABC is now named tox-user1"?

**Comment by sudden6** [comment]
___
*[src/widget/form/groupchatform.cpp, line 323 at r1](https://reviewable.io/reviews/qtox/qtox/5469#-LUNDSj5-n40wQVp6BQu:-LUNDSj5-n40wQVp6BQv:b8ewgl) ([raw file](https://github.com/qtox/qtox/blob/48d8b3b6ceaa72eb9e45ff82e6bec893e8bc4081/src/widget/form/groupchatform.cpp#L323)):*
> ```cpp
>     if (groupLast.isEmpty())
>     {
>         groupLast = group->getPeerList();
> ```

why get the peer list again here and not use `peers`?

**Review by sudden6 — COMMENTED** [review]
Your coding style doesn't match the qTox style, please reformat. you can use `tools/format-code.sh` and commit only the files you changed to fix that.

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5469#-:-LUNEQKf-et2EG7o7OjG:bfhz6m)* status: 0 of 1 LGTMs obtained

**Comment by Monsterovich** [comment]
>why get the peer list again here and not use peers?

What? I use them below in "for" loop.

**Comment by Monsterovich** [comment]
>Your coding style doesn't match the qTox style, please reformat. you can use tools/format-code.sh and commit only the files you changed to fix that.

Fixed it manually. Do not merge it. I have to fix one issue with nicknames.

**Comment by sudden6** [comment]
> What? I use them below in "for" loop.

A few lines above the for loop. You should probably take a look at the reviewable page (button in your first commit) to get the context for the comments.

**Comment by Monsterovich** [comment]
I have one issue: when user changes the nickname, it's not changed on other clients:

default


```
[12:56:53.494 UTC] core/core.cpp:508 : Debug: "Group 0 peerlist changed"
[12:56:53.495 UTC] widget/form/groupchatform.cpp:329 : Debug: "name: Monsterovich"
[12:56:53.495 UTC] widget/form/groupchatform.cpp:329 : Debug: "name: <Пусто>"
[12:56:53.544 UTC] core/core.cpp:516 : Debug: "Group 0, Peer 1, name changed to wow"
[12:56:53.546 UTC] widget/form/groupchatform.cpp:329 : Debug: "name: Monsterovich"
[12:56:53.546 UTC] widget/form/groupchatform.cpp:329 : Debug: "name: wow"
[12:57:18.544 UTC] core/core.cpp:516 : Debug: "Group 0, Peer 1, name changed to wow2" <<<<<<<
[12:57:18.546 UTC] widget/form/groupchatform.cpp:329 : Debug: "name: Monsterovich"
[12:57:18.546 UTC] widget/form/groupchatform.cpp:329 : Debug: "name: wow"
```

How can I fix this?

**Comment by sudden6** [comment]
Does `wow` have an alias set in your friends list? (Even if you didn't set one, verify this, there was a bug setting aliases, or try with two completely fresh profiles.)

Else I suspect a bug in this function or generally this file: https://github.com/qTox/qTox/blob/master/src/model/group.cpp#L45

**Comment by Monsterovich** [comment]
How can I refresh this list of users in group by group id? (actually refresh it in all groups)

I need to do this when:
- A local user has changed its nickname.
- A local user has set alias to someone.

The rest of the code works fine.

default

**Comment by Diadlo** [comment]
@Monsterovich It should be done automatically somehow like this: 

`Widget::createGroup` connects `Core::usernameSet` to `Group::setSelfName`. "local user has changed its nickname" -> `Core::setUsername` -> `emit Core::usernameSet` -> `Group::regeneratePeerList()` -> `emit Group::userListChanged`

Not sure, what is the best way to update names on alias change. `Friend::setAlias` -> `emit Friend::displayedNameChanged`, but who should regenerate group peer list? In our current bad architecture you can place it in `Widget` or `Nexus`, IDK. Just take all groups from `GroupList::getAllGroups` and regenerate all peer list. It's not quite efficient, but probably the best way for now

**Comment by Monsterovich** [comment]
Thanks to Diadlo for help.

Works as intended. Aliases work too. You can merge it. :)

**Comment by sudden6** [comment]
@Diadlo do you want to review and merge this PR?

File: src/widget/form/groupchatform.cpp [hunk 1]
```diff
@@ -307,6 +308,58 @@ void GroupChatForm::updateUserNames()
...
+                addSystemInfoMessage(tr("A new user has connected to the group"), ChatMessage::INFO, QDateTime::currentDateTime());
+            }
+            firstTime[peerPk] = true;
+            continue;
+        }
+        const QString name = FriendList::decideNickname(peerPk, peers.value(peerPk));
+        if (!groupLast.contains(peerPk)) {
+            groupLast.insert(peerPk, name);
+            addSystemInfoMessage(tr("%1 has joined the group").arg(name), ChatMessage::INFO, QDateTime::currentDateTime());
+        }
```
Comment (Diadlo) [inline src/widget/form/groupchatform.cpp]: Our code style says, that should be `} else {`

File: src/nexus.cpp [hunk 2]
```diff
@@ -213,6 +213,7 @@ void Nexus::showMainGUI()
     connect(core, &Core::friendTypingChanged, widget, &Widget::onFriendTypingChanged);
     connect(core, &Core::messageSentResult, widget, &Widget::onMessageSendResult);
     connect(core, &Core::groupSentFailed, widget, &Widget::onGroupSendFailed);
+    connect(core, &Core::usernameSet, widget, &Widget::refreshPeerListsLocal);
```
Comment (Diadlo) [inline src/nexus.cpp:216]: Just a note, not for this PR, but for The Plan™

Probably logic of groups update should be moved in the new `ContactManager` class. Or something like this.

**Review by Diadlo — CHANGES_REQUESTED** [review]

File: src/widget/form/groupchatform.cpp (hunk 1 above)
Comment (Monsterovich) [reply src/widget/form/groupchatform.cpp]: Done.

**Review by Diadlo — APPROVED** [review]
:lgtm_strong:

> Reviewed 1 of 2 files at r1, 8 of 9 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5469#-:-LUjJozY6JVCiRdILpu2:b-jbsf8z)* status: :shipit: complete! 1 of 1 LGTMs obtained

**Comment by Monsterovich** [comment]
Will you merge it?

**Comment by sudden6** [comment]
Done

**Pull request merged**